src/compiler/checker.ts

```diff
@@ -9401,7 +9401,30 @@ namespace ts {
                     error(node, Diagnostics.Value_of_type_0_is_not_callable_Did_you_mean_to_include_new, typeToString(funcType));
                 }
                 else {
-                    error(node, Diagnostics.Cannot_invoke_an_expression_whose_type_lacks_a_call_signature);
+                    if (apparentType.flags & TypeFlags.Union) {
+                        const unionApparentType = <UnionType>apparentType;
+                        let unionCallSignatures: Signature[] = [];
+                        let hasError = false;
+                        forEach(unionApparentType.types, type => {
+                            const eachCallSignature = getSignaturesOfType(type, SignatureKind.Call);
+                            if (!eachCallSignature.length) {
+                                error(node, Diagnostics.Cannot_invoke_an_expression_whose_type_lacks_a_call_signature_Type_0_has_no_compatible_call_signatures, typeToString(apparentType));

```

Comment:
You're now not giving any extra information based on the fact that you have a union type whose signatures weren't compatible. This different error message doesn't buy much, though it might be better in general if we used it all the time - I'd prefer you left that for a different PR.

You should create a separate diagnostic called `Not all types in union have compatible call signatures.`, chain the two with `createDiagnosticMessageChain` (`undefined` being a zero-list message chain) and use `createDiagnosticForNodeFromMessageChain`